README.md
# Ray Tracing in One Weekend 

[![Haskell CI](https://github.com/zarak/ray-tracing-one-weekend/actions/workflows/ray-tracing.yaml/badge.svg)](https://github.com/zarak/ray-tracing-one-weekend/actions/workflows/ray-tracing.yaml)

This repository is based on the book [_Ray Tracing in One Weekend_](https://raytracing.github.io/books/RayTracingInOneWeekend.html) by Peter Shirley. The code is my own implementation in Haskell, whereas the book uses C++.

![Final render](images/final.png)

### Commands
```
$ cabal run ray-tracing-one-weekend -- --help

Usage: ray-tracing-one-weekend [--imageWidth INT] [--samplesPerPixel INT]

  Ray tracer

Available options:
  --imageWidth INT         Set the image width
  --samplesPerPixel INT    Set the samples per pixel
  -h,--help                Show this help text
```


CHANGELOG.md
# Revision history for ray-tracing-one-weekend

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.

## 0.1.0.1 
* Optimizations and profiling

Parameters
```haskell
samplesPerPixel :: Int
samplesPerPixel = 100

aspectRatio :: Double
aspectRatio = 16.0 / 9.0

imageWidth :: Int
imageWidth = 40
```

Flags
```
cabal run ray-tracing-one-weekend --ghc-options="-O2 -fprof-auto" --enable-profiling -- +RTS -P
```

### Inline pragma for vector subtraction
```haskell
  v - w = Vec3 (v.x - w.x) (v.y - w.y) (v.z - w.z)
  {-# INLINE (-) #-}
```

Before
```
	Sat Jun 24 14:52 2023 Time and Allocation Profiling Report  (Final)

	   ray-tracing-one-weekend +RTS -P -RTS

	total time  =       50.54 secs   (50542 ticks @ 1000 us, 1 processor)
	total alloc = 58,108,777,712 bytes  (excludes profiling overheads)

COST CENTRE        MODULE    SRC                              %time %alloc  ticks     bytes

hit                Sphere    src/Sphere.hs:(17,3)-(34,103)     32.8    0.1  16558  57025848
-                  Vec3      src/Vec3.hs:27:3-50               16.0    9.7   8091 5634053728
hit.oc             Sphere    src/Sphere.hs:21:9-39             11.3   29.0   5731 16847195904
hit                Sphere    src/Sphere.hs:(51,3)-(64,54)       8.6   21.4   4354 12421111768
findNearestRoot    Sphere    src/Sphere.hs:(39,1)-(46,47)       7.9    0.0   3994   6912240
lengthSquared      Vec3      src/Vec3.hs:57:1-51                5.1    9.7   2564 5628614384
hit.c              Sphere    src/Sphere.hs:24:9-60              4.4    9.7   2249 5615731968
hit.a              Sphere    src/Sphere.hs:22:9-43              2.4    3.2   1234 1871910656
hit.halfB          Sphere    src/Sphere.hs:23:9-36              2.1    3.2   1051 1871910656
randomInUnitSphere RtWeekend src/RtWeekend.hs:(46,1)-(50,15)    1.5    2.9    750 1694182016
hit.sqrtd          Sphere    src/Sphere.hs:26:9-33              1.4    0.0    699         0
randomDoubleR      RtWeekend src/RtWeekend.hs:(34,1)-(38,23)    1.3    1.6    665 956636776
lambertian         Material  src/Material.hs:(12,1)-(22,19)     1.2    2.0    595 1151823304
mkWorld            MyLib     src/MyLib.hs:118:1-34              1.0    1.8    490 1022208000
randomScene        MyLib     src/MyLib.hs:(121,1)-(158,65)      0.2    2.3     85 1351965080
```

After 
```
	Sat Jun 24 14:55 2023 Time and Allocation Profiling Report  (Final)

	   ray-tracing-one-weekend +RTS -P -RTS

	total time  =       52.58 secs   (52580 ticks @ 1000 us, 1 processor)
	total alloc = 37,496,581,696 bytes  (excludes profiling overheads)

COST CENTRE        MODULE    SRC                              %time %alloc  ticks     bytes

hit                Sphere    src/Sphere.hs:(17,3)-(34,103)     52.0    0.2  27360  57025864
-                  Vec3      src/Vec3.hs:27:3-50                9.0   15.0   4714 5634053728
findNearestRoot    Sphere    src/Sphere.hs:(39,1)-(46,47)       7.3    0.0   3858   6912240
hit.oc             Sphere    src/Sphere.hs:21:9-39              7.1   39.9   3753 14975285248
lengthSquared      Vec3      src/Vec3.hs:58:1-51                5.0    5.0   2639 1884793072
hit.c              Sphere    src/Sphere.hs:24:9-60              4.2    5.0   2187 1871910656
hit.a              Sphere    src/Sphere.hs:22:9-43              2.8    5.0   1454 1871910656
hit                Sphere    src/Sphere.hs:(51,3)-(64,54)       2.3    3.2   1215 1189647816
hit.halfB          Sphere    src/Sphere.hs:23:9-36              2.0    5.0   1059 1871910656
randomInUnitSphere RtWeekend src/RtWeekend.hs:(46,1)-(50,15)    1.6    4.5    867 1694182016
lambertian         Material  src/Material.hs:(12,1)-(22,19)     1.2    3.1    621 1151823304
randomDoubleR      RtWeekend src/RtWeekend.hs:(34,1)-(38,23)    1.2    2.6    615 956636776
mkWorld            MyLib     src/MyLib.hs:118:1-34              0.9    2.7    476 1022208000
uniformRM          Vec3      src/Vec3.hs:(39,3)-(43,21)         0.8    1.4    412 533407816
metal              Material  src/Material.hs:(25,1)-(34,19)     0.5    1.5    272 571762984
randomScene        MyLib     src/MyLib.hs:(121,1)-(158,65)      0.2    3.6     97 1351965112
```


### Strictness annotations 


```haskell
data Vec3 = Vec3
  { x :: !Double,
    y :: !Double,
    z :: !Double
  }
  deriving (Show, Eq)

data Sphere = Sphere
  { center :: !Point,
    radius :: !Double,
    material :: !Material
  }
```

Result
```
	Sat Jun 24 15:23 2023 Time and Allocation Profiling Report  (Final)

	   ray-tracing-one-weekend +RTS -P -RTS

	total time  =       60.92 secs   (60922 ticks @ 1000 us, 1 processor)
	total alloc = 44,468,304,240 bytes  (excludes profiling overheads)

COST CENTRE        MODULE    SRC                              %time %alloc  ticks     bytes

hit                Sphere    src/Sphere.hs:(17,3)-(42,13)      45.4    0.1  27658  57025864
-                  Vec3      src/Vec3.hs:36:3-50               13.6   12.7   8255 5634053728
hit.oc             Sphere    src/Sphere.hs:21:9-39              7.5   37.9   4556 16847195904
findNearestRoot    Sphere    src/Sphere.hs:(47,1)-(54,47)       6.8    0.0   4122   6912240
lengthSquared      Vec3      src/Vec3.hs:66:1-51                4.0   12.7   2434 5628614384
hit.c              Sphere    src/Sphere.hs:24:9-60              3.8    4.2   2295 1871910656
hit.a              Sphere    src/Sphere.hs:22:9-43              3.1    4.2   1862 1871910656
hit                Sphere    src/Sphere.hs:(59,3)-(72,54)       2.2    2.7   1313 1189647816
$sel:center:Sphere Sphere    src/Sphere.hs:11:5-10              2.1    0.0   1272         0
hit.halfB          Sphere    src/Sphere.hs:23:9-36              2.0    4.2   1224 1871910656
base               Ray       src/Ray.hs:6:5-8                   1.6    0.0    949         0
randomInUnitSphere RtWeekend src/RtWeekend.hs:(46,1)-(50,15)    1.3    3.8    774 1694182016
randomDoubleR      RtWeekend src/RtWeekend.hs:(34,1)-(38,23)    1.1    2.2    688 956636776

src/MyLib.hs
{-# LANGUAGE MultiWayIf #-}
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}

{-# HLINT ignore "Use fold" #-}
{-# HLINT ignore "Redundant pure" #-}
module MyLib (someFunc, rayColor, generateLine, randomScene) where

import Camera
import Color (Color (..), color, scaleColor, white, writeColor)
import Control.Concurrent.Async (mapConcurrently)
import Control.Monad (replicateM)
import Data.Maybe (catMaybes)
import Data.Text qualified as T
import Data.Text.IO qualified as T (putStrLn)
import GHC.Real (infinity)
import Hittable
import Material
import MaterialHit
import MovingSphere (MovingSphere (..))
import Ray
import RtWeekend
import Sphere
import System.IO (hFlush, hPutStr, stderr)
import System.Random.MWC
import System.Random.MWC qualified as MWC
import Text.Printf
import Vec3

samplesPerPixel :: Int
samplesPerPixel = 100

shadowAcne :: Double
shadowAcne = 0.001

aspectRatio :: Double
aspectRatio = 16.0 / 9.0

imageWidth :: Int
imageWidth = 400

imageHeight :: Int
imageHeight = truncate $ fromIntegral imageWidth / aspectRatio

camera :: Camera
-- camera = mkCamera (point -2 2 1) (point 0 0 -1) (Vec3 0 1 0) 90.0 aspectRatio
camera =
  mkCamera
    lookfrom
    lookat
    vup
    20
    aspectRatio
    aperture
    distToFocus
    time0
    time1
  where
    lookfrom = point 13 2 3
    lookat = Point zeros
    vup = Vec3 0 1 0
    distToFocus = 10.0
    aperture = 0.1
    time0 = 0.0
    time1 = 1.0

rayColor :: Ray -> GenIO -> Int -> World AnyHittable -> IO Color
rayColor _ _ 0 _ = pure mempty
rayColor r g depth world = do
  let unitDirection = unitVector r.direction
      t = 0.5 * (unitDirection.y + 1.0)
      blue = color 0.5 0.7 1.0
      blueWhiteLerp = scaleColor (1.0 - t) white <> scaleColor t blue

  case hit world r (shadowAcne, fromRational infinity) of
    Nothing -> pure blueWhiteLerp
    Just rec -> do
      mScattered <- rec.material.scatter g r rec
      case mScattered of
        Nothing -> pure mempty
        Just scattered -> do
          c <- rayColor scattered.ray g (depth - 1) world
          pure $ Color $ scattered.attenuation.toVec3 * c.toVec3

drawRay :: Int -> Int -> World AnyHittable -> IO Color
drawRay i j world = do
  g <- MWC.create
  x <- randomDouble g
  y <- randomDouble g
  let u = (fromIntegral i + x) / fromIntegral (imageWidth - 1)
      v = (fromIntegral j + y) / fromIntegral (imageHeight - 1)
  r <- cameraRay camera u v g
  pixelColor <- rayColor r g maximumDepth world
  pure pixelColor

generateLine :: Int -> World AnyHittable -> IO ()
generateLine j world = do
  sampledColors <-
    mapConcurrently
      ( \i -> do
          cs <- replicateM samplesPerPixel $ drawRay i j world
          let summedColors = foldr (<>) mempty cs
          pure $ writeColor summedColors samplesPerPixel
      )
      [1 .. imageWidth]
  T.putStrLn $ T.unlines sampledColors

generateImage :: Int -> World AnyHittable -> IO ()
generateImage 0 _ = do
  hPutStr stderr "\rScanlines remaining: 0 \nDone.\n"
  hFlush stderr
generateImage j world = do
  hPutStr stderr ("\rScanlines remaining: " ++ show j ++ "\n")
  hFlush stderr
  generateLine j world
  generateImage (j - 1) world

someFunc :: IO ()
someFunc = do
  -- g <- MWC.createSystemRandom
  g <- MWC.create -- use for testing
  putStrLn $ printf "P3\n%d %d\n255" imageWidth imageHeight
  world <- World <$> randomScene g
  generateImage imageHeight world

randomScene :: GenIO -> IO [AnyHittable]
randomScene g = do
  let groundMaterial = lambertian (color 0.5 0.5 0.5)
      largeSphere = AnyHittable (Sphere (point 0 -1000 0) 1000 groundMaterial)

  smallSpheres <-
    fmap catMaybes . sequence $
      ( [ do
            chooseMat <- randomDouble g
            -- Generate the center point of each sphere
            let center =
                  point
                    (fromIntegral a + 0.9 * chooseMat)
                    0.2 -- y-value is fixed so that spheres are all at ground level (the same height)
                    (fromIntegral b + 0.9 * chooseMat)
            -- Proceed only if the small sphere is a certain distance away from
            -- one of the larger spheres, otherwise skip.
            if Vec3.length (point 4 0.2 0 |-> center) > 0.9
              then do
                if
                    -- Choose lambertian 80% of the time
                    | chooseMat < 0.8 -> createMovingAlbertian center g
                    -- Chose metal 15% of the time
                    | chooseMat < 0.95 -> createStaticMetal center g
                    -- Choose dieletric 5% of the time
                    | otherwise -> pure $ Just (AnyHittable $ Sphere center 0.2 (dielectric 1.5))
              else pure Nothing
          | -- Produce grid of small spheres
            a <- [-11 .. 10 :: Int],
            b <- [-11 .. 10 :: Int]
        ]
      )

  -- Three large spheres
  let material1 = dielectric 1.5
      sphere1 = AnyHittable $ Sphere (point 0 1 0) 1.0 material1
  let material2 = lambertian (color 0.4 0.2 0.1)
      sphere2 = AnyHittable $ Sphere (point -4 1 0) 1.0 material2
  let material3 = metal (color 0.7 0.6 0.5) 0.0
      sphere3 = AnyHittable $ Sphere (point 4 1 0) 1.0 material3

  pure $ largeSphere : sphere1 : sphere2 : sphere3 : smallSpheres

createMovingAlbertian :: Point -> GenIO -> IO (Maybe AnyHittable)
createMovingAlbertian center g = do
  albedo <- (*) <$> uniformM g <*> uniformM g
  let material = lambertian (Color albedo)
  y <- randomDoubleR 0 0.5 g
  let center2 = Point (Vec3 0 y 0 + center.toVec3)
  pure $ Just (AnyHittable $ MovingSphere center center2 0.0 1.0 0.2 material)

createStaticMetal :: Point -> GenIO -> IO (Maybe AnyHittable)
createStaticMetal center g = do
  albedo <- Color <$> uniformRM (0.5, 1) g
  fuzz <- randomDoubleR 0 0.5 g
  pure $ Just (AnyHittable $ Sphere center 0.2 (metal albedo fuzz))
